lib/maths/CBoostedTreeFactory.cc

```diff
@@ -0,0 +1,340 @@
+            m_Tree->m_Impl->m_Lambda = m_Tree->m_Impl->m_GammaOverride ? lambda : 0.5 * lambda;
+        }
+        if (m_Tree->m_Impl->m_GammaOverride == boost::none) {
+            m_Tree->m_Impl->m_Gamma = m_Tree->m_Impl->m_LambdaOverride ? gamma : 0.5 * gamma;
+        }
+        LOG_TRACE(<< "lambda(initial) = " << m_Tree->m_Impl->m_Lambda
+                  << " gamma(initial) = " << m_Tree->m_Impl->m_Gamma);
+    }
+
+    m_Tree->m_Impl->m_MaximumTreeSizeFraction = 10.0;
+
+    if (m_Tree->m_Impl->m_MaximumNumberTreesOverride == boost::none) {
+        // We allow a large number of trees by default in the main parameter
+        // optimisation loop. In practice, we should use many fewer if they
+        // don't significantly improve test error.
+        m_Tree->m_Impl->m_MaximumNumberTrees *= 10;
+    }
+}
+
+CBoostedTreeFactory
+CBoostedTreeFactory::constructFromParameters(std::size_t numberThreads,
+                                             std::size_t dependentVariable,
+                                             CBoostedTree::TLossFunctionUPtr loss) {
+    return {numberThreads, dependentVariable, std::move(loss)};
+}
+
+CBoostedTreeFactory::CBoostedTreeFactory(std::size_t numberThreads,
+                                         std::size_t dependentVariable,
+                                         CBoostedTree::TLossFunctionUPtr loss)
+    : m_Tree{std::make_unique<CBoostedTree>(numberThreads, dependentVariable, std::move(loss))} {
+}
+
+CBoostedTreeFactory& CBoostedTreeFactory::numberFolds(std::size_t folds) {
+    this->m_Tree->m_Impl->numberFolds(folds);
+    return *this;
+}
+
+CBoostedTreeFactory& CBoostedTreeFactory::lambda(double lambda) {
+    this->m_Tree->m_Impl->lambda(lambda);
+    return *this;
+}
+
+CBoostedTreeFactory& CBoostedTreeFactory::gamma(double gamma) {
+    this->m_Tree->m_Impl->gamma(gamma);
+    return *this;
+}
+
+CBoostedTreeFactory& CBoostedTreeFactory::eta(double eta) {
+    this->m_Tree->m_Impl->eta(eta);
+    return *this;
+}
+
+CBoostedTreeFactory& CBoostedTreeFactory::maximumNumberTrees(std::size_t maximumNumberTrees) {
+    this->m_Tree->m_Impl->maximumNumberTrees(maximumNumberTrees);
+    return *this;
+}
+
+CBoostedTreeFactory& CBoostedTreeFactory::featureBagFraction(double featureBagFraction) {
+    this->m_Tree->m_Impl->featureBagFraction(featureBagFraction);
+    return *this;
+}
+
+CBoostedTreeFactory&
+CBoostedTreeFactory::maximumOptimisationRoundsPerHyperparameter(std::size_t rounds) {
+    this->m_Tree->m_Impl->maximumOptimisationRoundsPerHyperparameter(rounds);
+    return *this;
+}
+
+CBoostedTreeFactory&
+CBoostedTreeFactory::progressCallback(CBoostedTree::TProgressCallback callback) {
+    this->m_ProgressCallback = callback;
+    return *this;
+}
+
+CBoostedTreeFactory& CBoostedTreeFactory::frame(core::CDataFrame& frame) {
+    this->m_Frame = &frame;
+    return *this;
+}
+CBoostedTreeFactory::operator CBoostedTree &&() {
+    return std::move(*(this->build()));
```

Comment:
I'm not sure about this: the unique pointer is moved out of build and then surely dies after this function returns. Even if it is moving does this guaranty that there isn't an invalid read: feels like a dark corner of the language. I think we can just stick with the version which returns a unique pointer can't we?

Reply:
Ok. I had to for a better integration with the client code. But if you don't mind, I leave only unique pointer option.

Reply:
Ok, no I don't mind if you need to change a little more calling code. I think we should stick with just the `unique_ptr` version.